messaging/schema.py

```diff
@@ -0,0 +1,243 @@
+from common.schema import LanguageEnum
+from common.utils import (
+    get_obj_if_user_can_administer,
+    project_user_required,
+    update_object_with_translations,
+)
+from events.models import Event, Occurrence
+from events.schema import OccurrenceNode
+from kukkuu.exceptions import DataValidationError, MessageAlreadySentError
+
+from .models import Message
+
+logger = logging.getLogger(__name__)
+
+MessageTranslation = apps.get_model("messaging", "MessageTranslation")
+
+
+class OccurrenceList(graphene.ObjectType):
+    occurrences = graphene.List(OccurrenceNode)
+
+
+class MessageTranslationType(DjangoObjectType):
+    class Meta:
+        model = MessageTranslation
+        exclude = ("id", "master")
+
+
+class RecipientSelectionEnum(graphene.Enum):
+    ALL = "all"
+    INVITED = "invited"
+    ENROLLED = "enrolled"
+    ATTENDED = "attended"
+    SUBSCRIBED_TO_FREE_SPOT_MESSAGE = "subscribed_to_free_spot_message"
+
+
+class MessageNode(DjangoObjectType):
+    subject = graphene.String()
+    body_text = graphene.String()
+    recipient_selection = graphene.Field(RecipientSelectionEnum)
+
+    class Meta:
+        model = Message
+        interfaces = (relay.Node,)
+        fields = (
+            "id",
+            "project",
+            "created_at",
+            "updated_at",
+            "subject",
+            "body_text",
+            "recipient_selection",
+            "event",
+            "occurrences",
+            "sent_at",
+            "recipient_count",
+            "translations",
+        )
+
+    @classmethod
+    def get_queryset(cls, queryset, info):
+        return queryset.prefetch_related("translations")
+
+
+class MessageTranslationsInput(graphene.InputObjectType):
+    language_code = LanguageEnum(required=True)
+    subject = graphene.String()
+    body_text = graphene.String()
+
+
+class AddMessageMutation(graphene.relay.ClientIDMutation):
+    class Input:
+        translations = graphene.List(MessageTranslationsInput)
+        project_id = graphene.GlobalID()
+        recipient_selection = RecipientSelectionEnum(required=True)
+        event_id = graphene.ID()
+        occurrence_ids = graphene.List(graphene.NonNull(graphene.ID))
+
+    message = graphene.Field(MessageNode)
+
+    @classmethod
+    @project_user_required
+    @transaction.atomic
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        data = deepcopy(kwargs)
+
+        data["project"] = get_obj_if_user_can_administer(
+            info, data.pop("project_id"), Project
+        )
+
+        if "event_id" in data:
+            data["event"] = get_obj_if_user_can_administer(
+                info, data.pop("event_id"), Event
+            )
+        else:
+            data["event"] = None
+        occurrences = [
+            get_obj_if_user_can_administer(info, occurrence_id, Occurrence)
+            for occurrence_id in data.pop("occurrence_ids", [])
+        ]
+
+        validate_event_and_occurrences(data["event"], occurrences)
```

Comment:
Will this work as expected if `data["event"]` is `None`?

Reply:
That should not be possible because of the `if` starting from line 100.

Reply:
Well I should have asked clearer but does line 116 set data["event"] to `None`? I meant what if data["event"] is `None` during the validation

Reply:
No, you shouldn't have, your question was great, but I was on crack. It should be fixed now.